diff --git a/gio/tests/gsubprocess-testprog.c b/gio/tests/gsubprocess-testprog.c
index b61ccc2..a7063b2 100644
--- a/gio/tests/gsubprocess-testprog.c
+++ b/gio/tests/gsubprocess-testprog.c
@@ -124,7 +124,7 @@ sleep_forever_mode (int argc,
   return 0;
 }
 
-#ifdef G_OS_UNIX
+#if defined(G_OS_UNIX) && !defined(__MVS__)
 #define GLIB_FD_CLOEXEC "e"
 #else
 #define GLIB_FD_CLOEXEC ""
diff --git a/gio/tests/iptosmessage.c b/gio/tests/iptosmessage.c
index ec7c092..28acdcb 100644
--- a/gio/tests/iptosmessage.c
+++ b/gio/tests/iptosmessage.c
@@ -26,7 +26,7 @@
 
 /* See the g_test_skip() calls below for platform-specific reasons why this test
  * code sometimes needs to be skipped. */
-#if ! (defined(G_OS_WIN32) || defined(__APPLE__) || defined(__GNU__) || defined(_AIX) || defined(__sun__))
+#if ! (defined(G_OS_WIN32) || defined(__APPLE__) || defined(__GNU__) || defined(_AIX) || defined(__sun__) || defined(__MVS__))
 
 static GSocketControlMessage *
 send_recv_control_message (GSocketFamily family, GSocketControlMessage *msg)
@@ -120,6 +120,8 @@ test_ip_tos (void)
   g_test_skip ("IP_RECVTOS not supported on AIX");
 #elif defined(__sun__)
   g_test_skip ("IP_RECVTOS not supported on Solaris");
+#elif defined(__MVS__)
+  g_test_skip ("IP_RECVTOS not supported on z/OS");
 #else
   GIPTosMessage *smsg;
   GIPTosMessage *rmsg;
@@ -148,6 +150,8 @@ test_ipv6_tclass (void)
   g_test_skip ("IPV6_RECVTCLASS not supported on AIX");
 #elif defined(__sun__)
   g_test_skip ("IPV6_RECVTCLASS not supported on Solaris");
+#elif defined(__MVS__)
+  g_test_skip ("IPV6_RECVTCLASS not supported on z/OS");
 #else
   GIPv6TclassMessage *smsg;
   GIPv6TclassMessage *rmsg;
