diff --git a/glib/gfileutils.c b/glib/gfileutils.c
index 37c2888..191a520 100644
--- a/glib/gfileutils.c
+++ b/glib/gfileutils.c
@@ -939,7 +939,7 @@ get_contents_posix (const gchar  *filename,
       FILE *f;
       gboolean retval;
 
-      f = fdopen (fd, "re");
+      f = fdopen (fd, "r");
       
       if (f == NULL)
         {
diff --git a/glib/glib-private.h b/glib/glib-private.h
index ac97142..0828810 100644
--- a/glib/glib-private.h
+++ b/glib/glib-private.h
@@ -64,7 +64,7 @@
  * as we'd like: https://stackoverflow.com/a/11529277/210151 and
  * https://devblogs.microsoft.com/oldnewthing/20200731-00/?p=104024
  */
-#elif defined (G_OS_UNIX) && !defined (__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && \
+#elif defined (G_OS_UNIX) && !defined (__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && !defined(__MVS__) && \
       g_macro__has_attribute (weak)
 
 #define HAS_DYNAMIC_ASAN_LOADING
diff --git a/glib/glib-unix.c b/glib/glib-unix.c
index 24a20df..c5eab8d 100644
--- a/glib/glib-unix.c
+++ b/glib/glib-unix.c
@@ -952,7 +952,7 @@ char *
 g_unix_fd_query_path (int      fd,
                       GError **error)
 {
-#if defined(__linux__) || defined(__sun) || defined(_AIX) || defined (__CYGWIN__)
+#if defined(__linux__) || defined(__sun) || defined(__MVS__) || defined(_AIX) || defined (__CYGWIN__)
   char *path;
   char *proc_path;
 
diff --git a/glib/gutils.c b/glib/gutils.c
index 42eb9fe..ade6e1b 100644
--- a/glib/gutils.c
+++ b/glib/gutils.c
@@ -734,7 +734,7 @@ g_get_user_database_entry (void)
           {
             e.user_name = g_strdup (pw->pw_name);
 
-#ifndef __BIONIC__
+#if !defined(__BIONIC__) && !defined(__MVS__)
             if (pw->pw_gecos && *pw->pw_gecos != '\0' && pw->pw_name)
               {
                 gchar **gecos_fields;
