diff --git a/gobject/meson.build b/gobject/meson.build
index 9690e7c..7027405 100644
--- a/gobject/meson.build
+++ b/gobject/meson.build
@@ -124,23 +124,23 @@ glib_enumtypes_input_headers = files(
 
 glib_enumtypes_h = custom_target('glib_enumtypes_h',
   output : 'glib-enumtypes.h',
-  capture : true,
+  capture : false,
   input : glib_enumtypes_input_headers,
   install : true,
   install_dir : join_paths(get_option('includedir'), 'glib-2.0/gobject'),
   install_tag: 'devel',
   command : [python, glib_mkenums,
              '--template', files('glib-enumtypes.h.template'),
-             '@INPUT@'])
+             '--output', '@OUTPUT@', '@INPUT@'])
 
 glib_enumtypes_c = custom_target('glib_enumtypes_c',
   output : 'glib-enumtypes.c',
-  capture : true,
+  capture : false,
   input : glib_enumtypes_input_headers,
   depends : [glib_enumtypes_h],
   command : [python, glib_mkenums,
              '--template', files('glib-enumtypes.c.template'),
-             '@INPUT@'])
+             '--output', '@OUTPUT@', '@INPUT@'])
 
 # Expose as variable to be used by gobject-introspection
 # when it includes GLib as a subproject
